### PR TITLE
Harmonize `FillValue` and `missing_value` during encoding and decoding steps

### PR DESCRIPTION
**Pull request by andersy005**
As pointed out in https://github.com/jbusecke/cmip6_preprocessing/issues/5, xarray appears to be very strict during the *encoding* and *decoding* steps even when there are (harmless) discrepancies between `missing_value` and `_FillValue`. For instance, when dtypes of `missing_value` and `_FillValue` are different, xarray gives up: 

```python
In [74]: from xarray.coding import variables                                                    

In [75]: import numpy as np                                                                     

In [76]: import xarray as xr                                                                    

In [77]: original = xr.Variable( 
    ...:         ("x",), 
    ...:         [0.0, -1.0, 1.0], 
    ...:         encoding={"_FillValue": np.float32(1e20), "missing_value": np.float64(1e20)}, 
    ...:     )                                                                                  

In [78]: coder = variables.CFMaskCoder()                                                        

In [79]: encoded = coder.encode(original)                                                       
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-79-9fbc3632e28b> in <module>
----> 1 encoded = coder.encode(original)

/glade/work/abanihi/devel/pangeo/xarray/xarray/coding/variables.py in encode(self, variable, name)
    156             raise ValueError(
    157                 "Variable {!r} has multiple fill values {}. "
--> 158                 "Cannot encode data. ".format(name, [fv, mv])
    159             )
    160 

ValueError: Variable None has multiple fill values [1e+20, 1e+20]. Cannot encode data. 
```

 - [ ] Closes #xxxx
 - [x] Tests added
 - [x] Passes `black . && mypy . && flake8`
 - [x] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API

### CODE REVIEW

**Comment by andersy005** [comment]
@dcherian, what is the right way to do the type casting in `encode()`?

I thought of trying something along these lines:

```python
encoding["missing_value"] = encoding["missing_value"].astype(data.dtype)
```
However, I quickly realized that this breaks when `encoding["missing_value"]` is not of a numpy object.

**Comment by dcherian** [comment]
Thanks for taking this on, Anderson.

You'll have to convert back from the numpy array before exiting that function. It may be simpler if you avoid the cast and just keep the change from `equivalent` to `allclose_or_equiv`.

For reference, here's how xarray determines `dtype` for netCDF4
https://github.com/pydata/xarray/blob/4e9240a2087ffbf119919e1ac98046bbf164f94d/xarray/backends/netCDF4_.py#L131-L141

(I found this by working looking for the `createVariable` statement here)
https://github.com/pydata/xarray/blob/4e9240a2087ffbf119919e1ac98046bbf164f94d/xarray/backends/netCDF4_.py#L439-L485

File: xarray/coding/variables.py [hunk 1]
```diff
@@ -152,18 +151,28 @@ def encode(self, variable, name=None):
...
+            and mv is not None
+            and not duck_array_ops.allclose_or_equiv(fv, mv)
+        ):
             raise ValueError(
                 "Variable {!r} has multiple fill values {}. "
                 "Cannot encode data. ".format(name, [fv, mv])
             )
 
         if fv is not None:
+            encoding["_FillValue"] = np.asarray(encoding["_FillValue"]).astype(
```
Comment (shoyer) [inline xarray/coding/variables.py]: Maybe using a NumPy scalar would work here? e.g., `encoding["_FillValue"] = data.dtype.type(encoding["_FillValue"])`

Actually I think you'll need to modify the `pop_to` line below, which overrides `_FillValue` in encoding.

Comment (andersy005) [reply xarray/coding/variables.py]: Thank you for the hint, @shoyer! 

> Actually I think you'll need to modify the pop_to line below, which overrides _FillValue in encoding.

I took a look at `pop_to()` function, and it appears that it is meant to just move a key/value from one dict to another dict. If I understand this correctly, we need to do type casting prior the `pop_to()` call, right?

**Comment by andersy005** [comment]
@dcherian, 

> You'll have to convert back from the numpy array before exiting that function.

I was able to address this using @shoyer's suggestion above.

File: xarray/coding/variables.py [hunk 2]
```diff
@@ -152,18 +151,26 @@ def encode(self, variable, name=None):
...
         mv = encoding.get("missing_value")
 
-        if fv is not None and mv is not None and not equivalent(fv, mv):
+        if (
+            fv is not None
+            and mv is not None
+            and not duck_array_ops.allclose_or_equiv(fv, mv)
+        ):
             raise ValueError(
                 "Variable {!r} has multiple fill values {}. "
```
Comment (dcherian) [inline xarray/coding/variables.py]: Let's make this nicer. Maybe something like 
```
f"Variable {name!r} has conflicting _FillValue ({fv}) and missing_value ({mv}). Cannot encode data."
```

File: xarray/coding/variables.py [hunk 3]
```diff
@@ -152,18 +151,26 @@ def encode(self, variable, name=None):
...
                 "Cannot encode data. ".format(name, [fv, mv])
             )
 
         if fv is not None:
+            # Ensure _FillValue is cast to same dtype as data's
+            encoding["_FillValue"] = data.dtype.type(encoding["_FillValue"])
```
Comment (dcherian) [inline xarray/coding/variables.py]: nit:
```suggestion
            encoding["_FillValue"] = data.dtype.type(fv)
```

File: xarray/coding/variables.py [hunk 4]
```diff
@@ -152,18 +151,26 @@ def encode(self, variable, name=None):
...
             fill_value = pop_to(encoding, attrs, "_FillValue", name=name)
             if not pd.isnull(fill_value):
                 data = duck_array_ops.fillna(data, fill_value)
 
         if mv is not None:
+            # Ensure missing_value is cast to same dtype as data's
+            encoding["missing_value"] = data.dtype.type(encoding["missing_value"])
```
Comment (dcherian) [inline xarray/coding/variables.py]: nit:
```suggestion
            encoding["missing_value"] = data.dtype.type(mv)
```

**Comment by andersy005** [comment]
@dcherian & @shoyer, thank you both for your help! 

In which section(bug fixes? enhancements?) in `whats-new.rst` should I document these changes?

**Comment by dcherian** [comment]
Let's do bug fixes

**Comment by max-sixty** [comment]
Thanks @andersy005 !

**Pull request merged**